docs/dev-docs/flow-control.md

```diff
@@ -41,19 +45,111 @@ The best way to handle these would be using states and state-management. But in
+
+2. Use the ERMrestJS's activeList API to generate the secondary requests for the page. The following is how this API works:
+
+    1. Since we want to allow users to open the citation popup as soon as possible, we're going to first send all the requests needed for it. So all the pseudo-columns listed in the `wait_for` of citation annotation will be sent first.
+
+    2. We want to show the main section of record app from top to bottom to avoid the jumps as much as possible. But since we can assume that the aggregates are slower than entity set and inline entity requests, we will:
+
+      2.1. Traverse the list of visible columns from top to bottom and generate the requests for columns that only have entity set in their `wait_for` or inline columns without any `wait_for`.
+
+      2.2. Traverse the list of visible columns the second time from top to bottom and if the column is aggregate, or has aggregate in its `wait_for`; generate requests for them (and their `wait_for` list definition).
+
+    3. For the related section, since we're showing the tables from top to bottom, we don't need to do the optimization that we did for the main section. We can send the requests from top to bottom (and left to right of their `wait_for` list).
+
+    To better understand this, let's assume the following definition (`i` is used to indicate inline table, `es` for entity set, `agg` for aggregate, and `r` for related):
+
+    ```json
+    "tag:isrd.isi.edu,2016:visible-columns": {
+        "detailed": [
+            {"sourcekey": "col_1", "display": {"wait_for": ["agg_1"]}},
+            {"sourcekey": "i_1", "display": {"wait_for": ["agg_1", "agg_2"]}},
+            {"sourcekey": "i_2", "display": {"wait_for": ["es_1", "es_3", "agg_1", "agg_3"]}},
+            {"sourcekey": "col_2", "display": {"wait_for": ["es_4"]}},
+            {"sourcekey": "i_3", "display": {"wait_for": ["es_1", "es_2"]}},
+            {"sourcekey": "agg_4"},
+            {"sourcekey": "i_4"},
+        ]
+    },
+    "tag:isrd.isi.edu,2016:visible-foreign-keys": {
+        "detailed": [
+            {"sourcekey": "r_1", "display": {"wait_for": ["agg_4", "agg_5"]}},
+            {"sourcekey": "r_2", "display": {"wait_for": ["es_1", "es_5", "agg_6", "agg_7"]}},
+            {"sourcekey": "r_3", "display": {"wait_for": ["es_1", "es_6"]}},
+            {"sourcekey": "r_4"},
+        ]
+    }
+    ```
+
+    The following would be the order of sending the requests (paranthesis has been added for readibility purposes):
+
+    ```
+    (es_4), (i_3, es_1, es_2), (i_4), (agg_1), (i_1, agg_2), (i_2, es_3, agg_3), (agg_4)
+    (r_1, agg_4, agg_5), (r_2, es_1, es_5, agg_6, agg_7), (r_3, es_1, es_6), (r4)
+    ```
+
+
+
+3. Send requests for all the aggregate columns that are used in the tabular mode of inline tables from to bottom.
+
+4. Send requests for all the aggregate columns that are used in the tabular mode of inline tables from top to bottom.
```

Comment:
I think you repeated this line from the one above.

Reply:
I forgot to update the second one to say related tables (instead of inline tables). I'll fix this.